Fix gem for rails 7

piyushswain opened the pull request:
This PR aims to fix this gem for Rails 7 and latest Solidus.

- Fixed an issue with autoloading constants in initializers
- Fixed `coffee-rails` dependency issue
- Removed `omniauth-amazon` dependency (No longer being maintained)
- Removed `omniauth-twitter` dependecy (No longer being maintained)
- Install Browser tools for CircleCI to fix failing browser tests
- Updated README
- Other fixes for rails 7 compatibility

piyushswain committed:
Fix autoload errors during config initialization

This fixes an error when autoloading the Spree::SocialConfiguration constant during initialization.
It also moves the config out of the Spree::SocialConfig constant and into a method while maintaining backwards compatibility.

This commit was originally taken from this solution https://github.com/solidusio-contrib/solidus_social/pull/98/commits/28f8bb2eb4cf2a8359d83270e5830632b118b169

piyushswain committed:
Fix coffee-rails dependency issue

This commit fixes an error for a missing coffee-rails dependency

Error fixed by this commit:
cannot load such file -- coffee_script (LoadError)

waiting-for-dev commented on .circleci/config.yml:
```diff
@@ -5,7 +5,7 @@ orbs:
   # to change this configuration every time a Solidus version is released
   # or goes EOL.
-  solidusio_extensions: solidusio/extensions@volatile
+  solidusio_extensions: solidusio/extensions@0.2.27
```
Why do we need that?

piyushswain replied:
Removed the commit that changed this

waiting-for-dev commented on lib/solidus_social/config.rb:
```diff
@@ -0,0 +1,13 @@
+  end
+
+  def self.const_missing(name)
+    name == :SocialConfig ? social_config : super
```
Wouldn't it work to append

```ruby
::Spree::SocialConfig = Spree::SocialConfiguration.new
```

in `lib/solidus_social/social_configuration.rb`?

piyushswain replied:
Modified this here [4b71338](https://github.com/solidusio-contrib/solidus_social/commit/4b71338044bcb42b46f8e8db2064cf33bae63920)

waiting-for-dev commented on lib/solidus_social/social_configuration.rb:
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
```
I think that we need to remove `app/models/spree/social_configuration.rb` once we move it to `lib`.

piyushswain replied:
Removed this file here [6d93219](https://github.com/solidusio-contrib/solidus_social/commit/6d9321941a614f0dd7541e5e00060a02c01dc75b)

waiting-for-dev commented on solidus_social.gemspec:
```diff
@@ -39,6 +39,8 @@ Gem::Specification.new do |spec|
   spec.add_dependency 'solidus_auth_devise'
   spec.add_dependency 'solidus_core', ['>= 2.0.0', '< 4']
   spec.add_dependency 'solidus_support', '~> 0.5'
+  spec.add_dependency 'multi_json'
```
I can't see any reference to a `MultiJson` constant. Have we tried just removing the dependency?

piyushswain replied:
The install fails with the following error when `multi_json` is removed
```
cannot load such file -- multi_json (LoadError)
```

waiting-for-dev replied:
Could you provide the entire stack of the error? :pray:

piyushswain replied:
```
Traceback (most recent call last):
	42: from bin/rails:4:in `<main>'
	41: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bootsnap-1.12.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	40: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bootsnap-1.12.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	39: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/railties-7.0.3.1/lib/rails/commands.rb:18:in `<main>'
	38: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/railties-7.0.3.1/lib/rails/command.rb:48:in `invoke'
	37: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/railties-7.0.3.1/lib/rails/command/base.rb:87:in `perform'
	36: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/thor-1.2.1/lib/thor.rb:392:in `dispatch'
	35: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/thor-1.2.1/lib/thor/invocation.rb:127:in `invoke_command'
	34: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/thor-1.2.1/lib/thor/command.rb:27:in `run'
	33: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/railties-7.0.3.1/lib/rails/commands/console/console_command.rb:101:in `perform'
	32: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/railties-7.0.3.1/lib/rails/command/actions.rb:14:in `require_application_and_environment!'
	31: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/railties-7.0.3.1/lib/rails/command/actions.rb:22:in `require_application!'
	30: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bootsnap-1.12.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	29: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bootsnap-1.12.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	28: from /Users/piyushswain/nebulab/solidus_social_test_app/config/application.rb:7:in `<main>'
	27: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bundler-2.3.6/lib/bundler.rb:176:in `require'
	26: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bundler-2.3.6/lib/bundler/runtime.rb:44:in `require'
	25: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bundler-2.3.6/lib/bundler/runtime.rb:44:in `each'
	24: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bundler-2.3.6/lib/bundler/runtime.rb:55:in `block in require'
	23: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bundler-2.3.6/lib/bundler/runtime.rb:55:in `each'
	22: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bundler-2.3.6/lib/bundler/runtime.rb:60:in `block (2 levels) in require'
	21: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bootsnap-1.12.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	20: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bootsnap-1.12.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	19: from /Users/piyushswain/nebulab/solidus_social/lib/solidus_social.rb:8:in `<main>'
	18: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/polyglot-0.3.5/lib/polyglot.rb:65:in `require'
	17: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bootsnap-1.12.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	16: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bootsnap-1.12.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	15: from /Users/piyushswain/nebulab/solidus_social/lib/solidus_social/engine.rb:7:in `<main>'
	14: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/polyglot-0.3.5/lib/polyglot.rb:65:in `require'
	13: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bootsnap-1.12.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	12: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bootsnap-1.12.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	11: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/omniauth-amazon-1.0.1/lib/omniauth-amazon.rb:1:in `<main>'
	10: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/polyglot-0.3.5/lib/polyglot.rb:65:in `require'
	 9: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bootsnap-1.12.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	 8: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bootsnap-1.12.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	 7: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/omniauth-amazon-1.0.1/lib/omniauth/amazon.rb:2:in `<main>'
	 6: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/polyglot-0.3.5/lib/polyglot.rb:65:in `require'
	 5: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bootsnap-1.12.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	 4: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bootsnap-1.12.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	 3: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/omniauth-amazon-1.0.1/lib/omniauth/strategies/amazon.rb:2:in `<main>'
	 2: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/polyglot-0.3.5/lib/polyglot.rb:65:in `require'
	 1: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bootsnap-1.12.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:15:in `require'
/Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bootsnap-1.12.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:15:in `require': cannot load such file -- multi_json (LoadError)
```

waiting-for-dev replied:
I see. It's a problem with amazon's Omniauth integration. I created a [PR](https://github.com/wingrunr21/omniauth-amazon/pull/19), but it doesn't look to be very well maintained.

waiting-for-dev commented on solidus_social.gemspec:
```diff
@@ -39,6 +39,8 @@ Gem::Specification.new do |spec|
   spec.add_dependency 'solidus_core', ['>= 2.0.0', '< 4']
   spec.add_dependency 'solidus_support', '~> 0.5'
+  spec.add_dependency 'multi_json'
+  spec.add_dependency 'coffee-rails'
```
I can't see coffee files. Maybe there were in the past, but it's no longer the case. Does anything break if we remove the dependency?

piyushswain replied:
Yes
The install fails and gives the following error message
```
cannot load such file -- coffee_script (LoadError)
```

waiting-for-dev replied:
Probably we can remove this line:

https://github.com/solidusio-contrib/solidus_social/blob/d97cecd7c1f53093fe37d8f47b3ee41e9bbad7f8/lib/solidus_social/engine.rb#L9

piyushswain replied:
Yeah, this can probably be removed

piyushswain replied:
Removed and fixed the dependency issue here [382d2b](https://github.com/solidusio-contrib/solidus_social/commit/382d2b15f9afb1843bccffb40dbe983539083118)

waiting-for-dev reviewed (changes requested):
Thanks, @piyushswain!!

I left some questions.

Also, would you mind fixing CI as part of this PR? If we need to release a new version so that `solidus_bolt` can pull it, we need it to be green.

waiting-for-dev commented on lib/solidus_social/config.rb:
```diff
@@ -0,0 +1,7 @@
+
+require 'solidus_social/social_configuration'
+
+module Spree
```
No need to wrap within the `Spree` module. I'd just add the `::Spree::SocialConfig` assignment in the `social_configuration.rb` file.

piyushswain replied:
Modified this here [36f081b](https://github.com/solidusio-contrib/solidus_social/commit/36f081b55005ad895a564c8c2cf34e63cb63b325)

Fixed now

waiting-for-dev commented:
Ref: https://github.com/solidusio-contrib/solidus_social/pull/102#discussion_r938726743

I'd remove both [omniauth-amazon](https://github.com/wingrunr21/omniauth-amazon) & [omniauth-twitter](https://github.com/arunagw/omniauth-twitter) from solidus_social, as they are the only integration that doesn't look like maintained anymore. Thoughts @aldesantis, @kennyadsl?

kennyadsl commented:
@waiting-for-dev that makes sense, especially if they are blocking us from upgrading this extension.

waiting-for-dev commented:
> @waiting-for-dev that makes sense, especially if they are blocking us from upgrading this extension.

Yeah, we just need to make sure we release a new major. @piyushswain, we can do it as part of this PR, if you like.

piyushswain commented:
> > @waiting-for-dev that makes sense, especially if they are blocking us from upgrading this extension.
> 
> Yeah, we just need to make sure we release a new major. @piyushswain, we can do it as part of this PR, if you like.

Okay, so I will remove `omniauth-twitter` and `omniauth-amazon` from the gem so that we can avoid the `multi-json` error.

piyushswain committed:
Remove omniauth-amazon

omniauth-amazon dependency has been removed since it is no longer being maintained.

piyushswain committed:
Remove omniauth-twitter

omniauth-twitter dependency has been removed since it is no longer being maintained.

waiting-for-dev commented:
@piyushswain, it looks like tests are failing

piyushswain commented:
@waiting-for-dev The tests pass on my development environment but a few tests fail on the CI with the following error
```
Webdrivers::BrowserNotFound:
            Failed to find Chrome binary.
```
All the failing tests have the same error.

I'm not sure if this is an issue in the CI or something to do with the gem itself.

waiting-for-dev commented:
> @waiting-for-dev The tests pass on my development environment but a few tests fail on the CI with the following error
> 
> ```
> Webdrivers::BrowserNotFound:
>             Failed to find Chrome binary.
> ```
> 
> All the failing tests have the same error.
> 
> I'm not sure if this is an issue in the CI or something to do with the gem itself.

@piyushswain, CircleCI images changed and now you need to require the browser tools manually. I think you only need to do the same that this commit:

https://github.com/solidusio/solidus_auth_devise/commit/e2503aa5b83c2448d686b3d7e8e7a87e1628877a

You can use [latest v1.4](https://circleci.com/developer/orbs/orb/circleci/browser-tools), though.

piyushswain committed:
Install Browser tools for CircleCI to fix failing browser tests

waiting-for-dev commented:
Cool! We've fixed the CircleCI issue. There're still some failures. Which version of Rails did you use locally for testing? If it was minor than 7, probably there's more to fix to make it compatible with Rails 7.

piyushswain commented:
> Cool! We've fixed the CircleCI issue. There're still some failures. Which version of Rails did you use locally for testing? If it was minor than 7, probably there's more to fix to make it compatible with Rails 7.

Tests run on the Solidus master branch on my local environment, so Rails 7
I am currently trying to fix the issues which are showing up on the CI

piyushswain commented:
> Cool! We've fixed the CircleCI issue. There're still some failures. Which version of Rails did you use locally for testing? If it was minor than 7, probably there's more to fix to make it compatible with Rails 7.

Hey @waiting-for-dev, I just noticed, the `sign_in_specs` related to facebook have been failing since today morning even on the `master` branch for solidus which were previously passing on local as well.
Investigating on the issue now.

piyushswain committed:
Fix failure on render file: method

Fixes the error where the render file: method fails to load a file mentioned without the complete path to the file

piyushswain committed:
Set the omniauth to version 1.9.1 in the gemspec

Needed to fix the version of omniauth to 1.9.1 as version >= 2.0 causes an error with omniauth-facebook where facebook sign in methods don't work.

piyushswain committed:
Fix failing feature test due to text mismatch

Fixed a failing test in features/spree/admin/authentication_methods_configuration_spec.rb which was failing due to a text mismatch.

waiting-for-dev commented on .circleci/config.yml:
```diff
@@ -1,6 +1,7 @@
 version: 2.1
 
 orbs:
+  browser-tools: circleci/browser-tools@1.1
```
```suggestion
  browser-tools: circleci/browser-tools@1.4
```

piyushswain replied:
Fixed here [d9fc568](https://github.com/solidusio-contrib/solidus_social/commit/d9fc5683dd3ca6ce0423e31baf2250a6f72fa1b1)

waiting-for-dev commented:
Thanks for your continuous work on it, @piyushswain. We're almost there! I left a suggestion, as we can update the browser tools versions.

Given that `omniauth-facebook` is forcing us to downgrade the support for `omniauth`, I'd also remove that provider. As we already support adding any provider manually, users can still easily do it by themselves. In fact, I think we should probably avoid adding any integration OOTB, as we're adding dependencies that some users won't need. But that's out of the scope of this PR.

Finally, we also need to update the README with the changes.

piyushswain commented:
> Thanks for your continuous work on it, @piyushswain. We're almost there! I left a suggestion, as we can update the browser tools versions.
> 
> Given that `omniauth-facebook` is forcing us to downgrade the support for `omniauth`, I'd also remove that provider. As we already support adding any provider manually, users can still easily do it by themselves. In fact, I think we should probably avoid adding any integration OOTB, as we're adding dependencies that some users won't need. But that's out of the scope of this PR.
> 
> Finally, we also need to update the README with the changes.

It's not that `omniauth-facebook` is forcing us to downgrade the `omniauth` version. The `omniauth` version was 1.9.1 before this PR as well.
`omniauth-facebook` isn't allowing us to update the `omniauth` version though.

But yes, we could remove it in another PR and users can add it later by themselves.

piyushswain committed:
README update for Rails 7 support
- Removed Mention of twitter support out of the box from the README
- Removed mention of amazon support out of the box from the README

waiting-for-dev commented:
> It's not that `omniauth-facebook` is forcing us to downgrade the `omniauth` version. The `omniauth` version was 1.9.1 before this PR as well. `omniauth-facebook` isn't allowing us to update the `omniauth` version though.
> 
> But yes, we could remove it in another PR and users can add it later by themselves.

Well, yeah, but I mean that it was open in the gemspec, and now we're requiring it to be lesser than v2. I'll definitely remove facebook instead of forcing users to use old versions of omniauth. It's not that they need to create a PR here, they can add facebook only in their apps if they need it. See https://github.com/solidusio-contrib/solidus_social#other-oauth-providers

DanielePalombo reviewed (approved):
LGTM.

Thank you @piyushswain. 
I have opened an issue to [remove omniauth-facebook](https://github.com/solidusio-contrib/solidus_social/issues/104).

DanielePalombo commented:
> > It's not that `omniauth-facebook` is forcing us to downgrade the `omniauth` version. The `omniauth` version was 1.9.1 before this PR as well. `omniauth-facebook` isn't allowing us to update the `omniauth` version though.
> > But yes, we could remove it in another PR and users can add it later by themselves.
> 
> Well, yeah, but I mean that it was open in the gemspec, and now we're requiring it to be lesser than v2. I'll definitely remove facebook instead of forcing users to use old versions of omniauth. It's not that they need to create a PR here, they can add facebook only in their apps if they need it. See https://github.com/solidusio-contrib/solidus_social#other-oauth-providers

@waiting-for-dev I have opened an [issue](https://github.com/solidusio-contrib/solidus_social/issues/104) to remove omniauth-facebook and the commit that locks the version. Is there something else to do with this PR? Otherwise, can you review it again so we can merge it? 🙏 

Thank you for your support.

waiting-for-dev commented:
Hey @DanielePalombo. No, we can merge it. But it'd be nice to remove `omniauth-facebook` before releasing the new version of it.
